src/rust/engine/process_execution/src/docker.rs

```diff
@@ -0,0 +1,330 @@
+use store::Store;
+use task_executor::Executor;
+use workunit_store::{in_workunit, RunningWorkunit};
+
+use crate::local::{
+  apply_chroot, create_sandbox, prepare_workdir, setup_run_sh_script, CapturedWorkdir, ChildOutput,
+  KeepSandboxes,
+};
+use crate::{
+  Context, FallibleProcessResultWithPlatform, ImmutableInputs, LocalCommandRunner, NamedCaches,
+  Platform, Process, ProcessError,
+};
+
+/// `CommandRunner` executes processes using a local Docker client.
+pub struct CommandRunner {
+  docker: Docker,
+  store: Store,
+  executor: Executor,
+  work_dir_base: PathBuf,
+  named_caches: NamedCaches,
+  immutable_inputs: ImmutableInputs,
+  keep_sandboxes: KeepSandboxes,
+  image: String,
+}
+
+impl CommandRunner {
+  pub fn new(
+    store: Store,
+    executor: Executor,
+    work_dir_base: PathBuf,
+    named_caches: NamedCaches,
+    immutable_inputs: ImmutableInputs,
+    keep_sandboxes: KeepSandboxes,
+    image: String,
+  ) -> Result<Self, String> {
+    let docker = Docker::connect_with_local_defaults()
+      .map_err(|err| format!("Failed to connect to local Docker: {err}"))?;
+    Ok(CommandRunner {
+      docker,
+      store,
+      executor,
+      work_dir_base,
+      named_caches,
+      immutable_inputs,
+      keep_sandboxes,
+      image,
+    })
+  }
+}
+
+#[async_trait]
+impl LocalCommandRunner for CommandRunner {
+  fn store(&self) -> &Store {
+    &self.store
+  }
+
+  fn named_caches(&self) -> &NamedCaches {
+    &self.named_caches
+  }
+
+  fn immutable_inputs(&self) -> &ImmutableInputs {
+    &self.immutable_inputs
+  }
+}
+
+impl fmt::Debug for CommandRunner {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    f.debug_struct("docker::CommandRunner")
+      .finish_non_exhaustive()
+  }
+}
+
+#[async_trait]
+impl super::CommandRunner for CommandRunner {
+  async fn run(
+    &self,
+    context: Context,
+    _workunit: &mut RunningWorkunit,
+    req: Process,
+  ) -> Result<FallibleProcessResultWithPlatform, ProcessError> {
+    let req_debug_repr = format!("{:#?}", req);
+    in_workunit!(
+      "run_local_process_via_docker",
+      req.level,
+      // NB: See engine::nodes::NodeKey::workunit_level for more information on why this workunit
+      // renders at the Process's level.
+      desc = Some(req.description.clone()),
+      |_workunit| async move {
+        let mut workdir = create_sandbox(
+          self.executor.clone(),
+          &self.work_dir_base,
+          &req.description,
+          self.keep_sandboxes,
+        )?;
+
+        // Start working on a mutable version of the process.
+        let mut req = req;
+
+        // Update env, replacing `{chroot}` placeholders with `/pants-work`. This is the mount point
+        // for the work directory within the Docker container.
+        //
+        // DOCKER-TODO: When dealing with invocations from cached containers, `{chroot}` should be
+        // replaced by the Pants executor process running inside the container.
```

Comment:
Interesting... I had been thinking that using `docker exec` would allow us to launch a new process inside an existing container, without needing a parent process inside the container.

Reply:
Yeah this comment is stale. The cached container approach will be spawn off containers using `--init` to add the "dummy" init process to reap child processes and keep the container alive. We can then `exec` into the container to run processes. (`bollard has a `create_exec` API to facilitate that.)

Longer term, a similar approach could be used to run processes remotely over SSH. Keep an SSH connection alive at the protocol level and then open channels for running processes as needed.